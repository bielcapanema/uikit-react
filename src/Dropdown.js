import cx from 'classnames'
import { Component, PropTypes, createElement } from 'react'

const links = {}

const subscribe = (namespace, listener) => {
  if (!links[namespace]) {
    links[namespace] = []
  }

  let isSubscribed = true

  links[namespace].push(listener)

  return function unsubscribe() {
    if (!isSubscribed) {
      return
    }

    isSubscribed = false

    const index = links[namespace].indexOf(listener)
    links[namespace].splice(index, 1)
  }
}
const dispatch = (namespace) => {
  const listeners = links[namespace]
  for (let i = 0; i < listeners.length; i++) { // eslint-disable-line no-plusplus
    const listener = listeners[i]
    listener()
  }
}

export default class Dropdown extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    delay: PropTypes.number.isRequired,
    mode: PropTypes.oneOf(['hover', 'click']).isRequired,
    remainTime: PropTypes.number.isRequired,
    component: PropTypes.node,
    dropdownLabel: PropTypes.string,
    link: PropTypes.string,
  }

  static defaultProps = {
    dropdownIcon: '',
    dropdownLabel: 'Open',
    mode: 'hover',
    remainTime: 800,
    delay: 0,
    component: 'div',
    link: '',
  }

  state = {
    isOpen: false,
  }

  componentDidMount() {
    if (this.props.link) {
      // listen to linked components
      this.unsubscribe = subscribe(this.props.link, () => {
        if (this.state.isOpen) {
          this.setState({ isOpen: false })
        }
      })
    }
  }

  componentWillUnmount() {
    if (this.leaveTimeout) {
      clearTimeout(this.leaveTimeout)
    }

    if (this.unsubscribe) {
      this.unsubscribe()
      this.unsubscribe = false
    }
  }

  handleMouseEnter = () => {
    if (this.leaveTimeout) {
      clearTimeout(this.leaveTimeout)
    }
    if (this.props.delay) {
      setTimeout(() => {
        this.setState({ isOpen: true })
      }, this.props.delay)
    } else {
      if (this.props.link) {
        dispatch(this.props.link)
      }
      this.setState({ isOpen: true })
    }

    if (this.props.mode === 'hover') {
      this.setState({ isOpen: true })
    }
  }

  handleMouseLeave = () => {
    if (this.props.remainTime) {
      this.leaveTimeout = setTimeout(() => {
        this.setState({ isOpen: false })
      }, this.props.remainTime)
    } else {
      this.setState({ isOpen: false })
    }
  }

  handleClick = () => {
    if (this.props.mode === 'click') {
      this.setState({ isOpen: !this.state.isOpen })
    }

    if (this.props.mode === 'hover') {
      this.setState({ isOpen: false })
    }
  }

  render() {
    const { handleMouseEnter, handleMouseLeave, handleClick } = this
    const { mode, children, component, link } = this.props
    const { isOpen } = this.state
    const className = cx('uk-button uk-button-default', {
      'uk-open': isOpen,
    })

    const dropdownClassName = cx('uk-dropdown uk-dropdown-bottom-left', {
      'uk-open': isOpen,
    })

    const labelProps = {
      'aria-expanded': isOpen,
      'aria-haspopup': true,
      className: link !== 'navbar' ? className : '',
      onClick: handleClick,
      onMouseEnter: mode === 'hover' && handleMouseEnter,
      onMouseLeave: mode === 'hover' && handleMouseLeave,
      children: this.props.dropdownLabel || mode,
    }

    const childrenProps = {
      'aria-expanded': isOpen,
      'aria-haspopup': true,
      className: dropdownClassName,
      onClick: handleClick,
      onMouseEnter: mode === 'hover' && handleMouseEnter,
      onMouseLeave: mode === 'hover' && handleMouseLeave,
      children,
    }

    const parentProps = {
      className: link !== 'navbar' ? 'uk-inline' : '',
      children: [
        createElement(link === 'navbar' ? 'a' : 'button', labelProps),
        createElement('div', childrenProps),
      ],
    }

    return (
      createElement(component, parentProps)
    )
  }
}
