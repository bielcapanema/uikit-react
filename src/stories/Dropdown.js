import { storiesOf } from '@kadira/storybook'

import Dropdown from '../Dropdown'

// This is to work around: https://github.com/kadirahq/react-storybook-addon-info/issues/26#issuecomment-229029177
Dropdown.displayName = 'Dropdown'

const Nav = () => <ul className="uk-nav uk-nav-dropdown">
  <li><a className="uk-warning" href="#">Item</a></li>
  <li><a href="#">Another item</a></li>
  <li className="uk-nav-header">Header</li>
  <li><a href="#">Item</a></li>
  <li><a href="#">Another item</a></li>
  <li className="uk-nav-divider" />
  <li><a href="#">Separated item</a></li>
</ul>

const Menu = () => <Nav />
Menu.displayName = 'Menu'

const DropdownMenu = () => <Nav />
DropdownMenu.displayName = 'Menu'

const LoremIpsum = `Lorem ipsum dolor sit amet, consectetur adipisicing elit,
  sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.`

storiesOf('Dropdown', module)
.addWithInfo('Basic Usage', '', () => (
  <div className="uk-margin-bottom">
    <Dropdown dropdownLabel="Hover" >
      {LoremIpsum}
    </Dropdown>
    <Dropdown dropdownLabel="Click" mode="click">
      {LoremIpsum}
    </Dropdown>
  </div>
), { header: false, inline: true, propTables: [Dropdown] })

  .addWithInfo('Navbar in Dropdown', '', () => (
    <div className="uk-margin-bottom">
      <Dropdown dropdownLabel="Hover">
        <Menu />
      </Dropdown>
      <Dropdown dropdownLabel="Click" mode="click">
        <Menu />
      </Dropdown>
    </div>
  ), { header: false, inline: true, propTables: [Dropdown] })

  .addWithInfo('Advanced', '', () => (
    <div className="uk-margin-bottom">
      <nav className="uk-navbar-container">
        <ul className="uk-navbar-nav">
          <li className="uk-active"><a>Active</a></li>
          <li><a href="">Item</a></li>
          <Dropdown
            className="uk-parent"
            component="li"
            dropdownLabel="Parent"
            link="navbar"
          >
            <DropdownMenu />
          </Dropdown>
          <Dropdown
            className="uk-parent"
            component="li"
            dropdownLabel="Parent"
            link="navbar"
          >
            <DropdownMenu />
          </Dropdown>
        </ul>

      </nav>
    </div>
  ), { header: false, inline: true })
