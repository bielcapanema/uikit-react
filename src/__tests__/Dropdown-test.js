import renderer from 'react-test-renderer'
import Dropdown from '../Dropdown'

jest.useFakeTimers()
const LoremIpsum = 'Lorem ipsum dolor sit amet, consectetur adipisicing elit'

describe('Dropdown', () => {
  it('renders correctly', () => {
    expect(renderer.create(
      <Dropdown dropdownLabel="Hover">
        {LoremIpsum}
      </Dropdown>
    ).toJSON()).toMatchSnapshot()

    expect(renderer.create(
      <Dropdown component="li" dropdownLabel="Parent" link="navbar">
        {LoremIpsum}
      </Dropdown>
    ).toJSON()).toMatchSnapshot()
  })

  it('changes the class when hovered', () => {
    let component = renderer.create(
      <Dropdown dropdownLabel="Hover">
        {LoremIpsum}
      </Dropdown>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback
    tree.children[1].props.onMouseEnter()
    expect(component.toJSON()).toMatchSnapshot()

    // manually trigger the callback
    tree.children[1].props.onMouseLeave()
    expect(component.toJSON()).toMatchSnapshot()

    // Fast forward timers, class should now be gone
    jest.runAllTimers()
    expect(component.toJSON()).toMatchSnapshot()

    let instance = component.getInstance()
    instance.handleMouseEnter()
    expect(instance.state.isOpen).toBeTruthy()

    component = renderer.create(
      <Dropdown delay={400} dropdownLabel="Hover" mode="hover" remainTime={0}>
        {LoremIpsum}
      </Dropdown>
    )
    instance = component.getInstance()
    instance.handleMouseEnter()
    expect(instance.state.isOpen).toBeTruthy()

    jest.runAllTimers()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    instance.handleMouseLeave()
    jest.runAllTimers()
    expect(tree).toMatchSnapshot()
    expect(instance.state.isOpen).toBe(false)
  })

  it('changes the class when clicked', () => {
    const component = renderer.create(
      <Dropdown dropdownLabel="Click" mode="click" >
        Click
      </Dropdown>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback
    tree.children[0].props.onClick()
    // re-rendering
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // manually trigger the callback
    tree.children[0].props.onClick()
    // re-rendering
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('closes sibling Dropdown on open', () => {
    const firstComponent = renderer.create(
      <Dropdown dropdownLabel="Hover" link="menu">
        First Instance
      </Dropdown>
    )
    let first = firstComponent.toJSON()
    expect(first).toMatchSnapshot()

    // manually trigger the callback
    first.children[1].props.onMouseEnter()

    first = firstComponent.toJSON()
    expect(first).toMatchSnapshot()
    expect(firstComponent.getInstance().state.isOpen).toBe(true)

    const secondComponent = renderer.create(
      <Dropdown dropdownLabel="Hover" link="menu">
        Second Instance
      </Dropdown>
    )
    let second = secondComponent.toJSON()
    expect(second).toMatchSnapshot()

    // manually trigger the callback
    second.children[1].props.onMouseEnter()

    // Verify that it is open
    second = secondComponent.toJSON()
    expect(second).toMatchSnapshot('should be open')
    expect(secondComponent.getInstance().state.isOpen).toBe(true)
    // Verify that the first component is closed
    first = firstComponent.toJSON()
    expect(first).toMatchSnapshot('should be closed')
    expect(firstComponent.getInstance().state.isOpen).toBe(false)

    // Verify there is a subscription
    const secondInstance = secondComponent.getInstance()
    expect({}.hasOwnProperty.call(secondInstance, 'unsubscribe')).toBe(true)
    const unsubscribe = secondInstance.unsubscribe

    // Start leave timeout before unmount
    second.children[1].props.onMouseLeave()

    // Trigger unmount
    secondComponent.update(<span />)
    // Should be no error from calling this again
    unsubscribe()
  })

  it('closes when clicked when in hover mode', () => {
    const component = renderer.create(
      <Dropdown dropdownLabel="Hover" mode="hover">
        Hover
      </Dropdown>
    )
    let tree = component.toJSON()

    // clicking the dropdown should not open it
    tree.children[0].props.onClick()
    tree = component.toJSON()
    expect(component.getInstance().state.isOpen).toBe(false)

    // open the dropdown
    tree.children[0].props.onMouseEnter()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component.getInstance().state.isOpen).toBe(true)

    // close the dropdown by click
    tree.children[0].props.onClick()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component.getInstance().state.isOpen).toBe(false)

    // clicking the dropdown again shouldn't open it
    tree.children[0].props.onClick()
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(component.getInstance().state.isOpen).toBe(false)
  })
})
