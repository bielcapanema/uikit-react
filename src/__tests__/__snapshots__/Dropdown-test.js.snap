// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown changes the class when clicked 1`] = `
<div
  className="uk-inline"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
  >
    Open
  </button>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-hidden"
    onClick={[Function]}
    onMouseEnter={false}
    onMouseLeave={false}
  >
    <button
      className="uk-button"
    >
      Click 
      <i
        className="uk-icon-caret-down"
      />
    </button>
  </div>
</div>
`;

exports[`Dropdown changes the class when clicked 2`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={false}
  onMouseLeave={false}
>
  <button
    className="uk-button"
  >
    Click 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown changes the class when clicked 3`] = `
<div
  aria-expanded={false}
  aria-haspopup={true}
  className="uk-inline"
  onClick={[Function]}
  onMouseEnter={false}
  onMouseLeave={false}
>
  <button
    className="uk-button"
  >
    Click 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown changes the class when hovered 1`] = `
<div
  className="uk-inline"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Open
  </button>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-hidden"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <button
      className="uk-button"
    >
      Hover 
      <i
        className="uk-icon-caret-down"
      />
    </button>
  </div>
</div>
`;

exports[`Dropdown changes the class when hovered 2`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown changes the class when hovered 3`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown changes the class when hovered 4`] = `
<div
  aria-expanded={false}
  aria-haspopup={true}
  className="uk-inline"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown changes the class when hovered 5`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown changes the class when hovered 6`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown closes sibling Dropdown on open 1`] = `
<div
  className="uk-inline"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Open
  </button>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-hidden"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <button
      className="uk-button"
    >
      First Instance 
      <i
        className="uk-icon-caret-down"
      />
    </button>
  </div>
</div>
`;

exports[`Dropdown closes sibling Dropdown on open 2`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    First Instance 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown closes sibling Dropdown on open 3`] = `
<div
  aria-expanded={false}
  aria-haspopup={true}
  className="uk-inline"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Second Instance 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown closes when clicked when in hover mode 1`] = `
<div
  aria-expanded={true}
  aria-haspopup={true}
  className="uk-inline uk-open"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown closes when clicked when in hover mode 2`] = `
<div
  aria-expanded={false}
  aria-haspopup={true}
  className="uk-inline"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown closes when clicked when in hover mode 3`] = `
<div
  aria-expanded={false}
  aria-haspopup={true}
  className="uk-inline"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    className="uk-button"
  >
    Hover 
    <i
      className="uk-icon-caret-down"
    />
  </button>
</div>
`;

exports[`Dropdown renders correctly 1`] = `
<div
  className="uk-inline"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Open
  </button>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-hidden"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <button
      className="uk-button"
    >
      Hover 
      <i
        className="uk-icon-caret-down"
      />
    </button>
  </div>
</div>
`;

exports[`Dropdown renders correctly 2`] = `
<li
  className="uk-inline"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-button uk-button-default"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Open
  </button>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    className="uk-hidden"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <a>
      Parent 
      <i
        className="uk-icon-caret-down"
      />
    </a>
  </div>
</li>
`;
